### ISSUE TITLE
Feature request: Allow "Custom configuration" for proxy and such, but require specific GUI access for this

### ISSUE THREAD

**Issue by sebastiannielsen** [opnsense/core#2589]
I have a feature request, and that is to add a "Custom configuration" option for all items including Proxy and such.

But, since this allow the GUI user to execute arbitary commands as discussed here:
https://forum.opnsense.org/index.php?topic=6516.0

I would suggest adding a extra option inside:
Effective Privileges

which is called: GUI --> Custom config

That is default off for all except the "System administrator" default account (even if you create a new account with Administrator privileges) and then you have to manually add that privilege to allow custom configs.

**Comment by AdSchellevis** [comment]
Sorry, we're not going to add custom entries in the configuration, administrators can deploy their own overwrites using the mechanisms in place:

https://github.com/opnsense/core/blob/master/src/opnsense/service/templates/OPNsense/Proxy/squid.conf#L281-L282

https://github.com/opnsense/core/blob/master/src/opnsense/service/templates/OPNsense/Proxy/squid.conf#L291-L292

**Comment by sebastiannielsen** [comment]
franco said it could be discussed, and I think its a easier solution if you could just change it from web interface.

By requiring a specific privilege that must manually added to user account, you avoid the main problem with "Arbitary code execution" as the privilege is default off for every user and must be enabled by system administrator.

For even more security, the privilege could be locked so only system administrator (the default account) can add that privilege, so even a user with privilege editing rights cannot add that privilege.

**Comment by AdSchellevis** [comment]
There are more problems with leaving arbitrary configuration entries in the configuration, they might break working setups after upgrades, leading to people complaining about issues which are close to impossible to catch. 

Personally, I wouldn't suggest adding custom text input to core, since it encourages user to copy-paste arbitrary data into their configuration which they might not always understand and is eventually impossible to validate using any model.

Usually the best approach is to functionally describe the feature you're seeking, sometimes it's rather easy to just add a checkbox or some input to help people out and make things accessible for the next use seeking something similar. During the course of time we've added more features for people.

In case you really need a plain squid.conf file for your application, there are two different options in my opinion:

1.  Use custom templates like described earlier
2. Build a plugin which injects your config text it into the configuration, specifically for your needs, which we might gain a spot in our plugins repository as long as it comes with a [very] big disclaimer.

**Comment by mimugmail** [comment]
Hm, does that mean I can include custom config related before and after authentication phase? Like custom regexp blacklists?

**Comment by fichtner** [comment]
Yes, flat files or generated content. We do have several web proxy plugins after all. ;)

**Comment by mimugmail** [comment]
I think I have a great idea <3
